KAFKA-7609: Add Protocol Generator for Kafka

cmccabe opened the pull request.

cmccabe commented:
Overview:

**buildSrc/src**
The message generator code is here.  This code is automatically re-run by gradle when one of the schema files changes.  The entire directory is processed at once to minimize the number of times we have to start a new JVM.  We use Jackson to translate the JSON files into Java objects.

**clients/src/main/java/org/apache/kafka/common/protocol/Message.java**
This is the interface implemented by all automatically generated messages.

**clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java**
Some utility functions used by the generated message code.

**clients/src/main/java/org/apache/kafka/common/protocol/Readable.java, Writable.java, ByteBufferAccessor.java**
The generated message code uses these classes for writing to a buffer.

**clients/src/main/message/README.md**
This README file explains how the JSON schemas work.

**clients/src/main/message/\*.json**
The JSON files in this directory implement every supported version of every Kafka API.  The unit tests automatically validate that the generated schemas match the hand-written schemas in our code.  Additionally, there are some things like request and response headers that have schemas here.

**clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java**
I added an optimization here for empty sets.  This is useful here because I want all messages to start with empty sets by default prior to being loaded with data.  This is similar to the "empty list" optimizations in the `java.util.ArrayList` class

**core/src, connect/src, clients/src/main/java/org/apache/kafka/clients/admin, etc.**
These changes are related to switching CreateTopicsRequest and CreateTopicsResponse to using the automatically generate classes rather than the hand-written ones.

stanislavkozlovski commented on buildSrc/src/main/java/org/apache/kafka/message/JsonUtil.java:
```diff
@@ -0,0 +1,75 @@
+        return JSON_SERDE.readValue(bld.toString(), clazz);
+    }
+
+    static String stripComments(String input) {
```
Should we have unit tests testing this functionality? Looks like a tricky function to me

cmccabe replied:
It turns out that Jackson has a built-in way to strip comments, so I just enabled that feature.

I added some unit tests for some of the other functions here.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java:
```diff
@@ -278,6 +278,16 @@ public ByteBuffer getBytes(String name) {
         return (ByteBuffer) result;
     }
 
+    public byte[] getByteArray(String name) {
```
Why do we need this?

cmccabe replied:
The generated code uses `byte[]` to represent fields with Bytes type.  The existing Struct code uses `ByteBuffer`.  That's why this function is needed now, whereas before we just had a function to get BBs.

This is kind of a debatable issue, I guess.  ByteBuffer is not more efficient (in fact, it's less efficient) unless you start sharing memory between multiple buffers, or using off-heap buffers, etc.  I can see a few places in the message processing code where we just turn around immediately and load / unload the BB to an array, which is just pure overhead.

On the other hand, perhaps it would be easier to integrate with the existing code if we provided ByteBuffer accessors and so forth?

The "records" fields are probably where we should get fancy with our optimizations.

Perhaps we should defer this to a follow-up discussion once we get the basic stuff in?

ijuma replied:
Yeah, I was thinking that it would be best to avoid changing things unless they were actually better. Using `byte[]` seems like going in the wrong direction even if in some cases we don't take advantage of using `ByteBuffer`.

cmccabe replied:
Perhaps we can revisit this in a follow-on JIRA?  I'm open to changing it, or even making it ByteBuffer everywhere, but for now byte[] is simpler and in some cases more efficient.

bob-barrett commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1347,59 +1348,47 @@ class KafkaApis(val requestChannel: RequestChannel,
+        topicResults += ((topic.name(), new ApiError(Errors.NOT_CONTROLLER, null)))
       }
-      sendResponseCallback(results)
+      sendResponseCallback(Map.empty)
```
nit: Modifying topicResults in the foreach and then passing Map.empty to the callback feels harder to understand to me than just passing the NOT_CONTROLLER result to callback. Is there a reason you want to do it this way?

cmccabe replied:
It is more for consistency with the code in the "else" block.

There might be a better way to structure, this though.  I have to say I would prefer just to have a map from topic names to results rather than dealing with all these lists.  But I didn't want to change AdminManager#createTopics.

cmccabe replied:
I have changed this in the latest version -- take a look.

bob-barrett commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1347,59 +1348,47 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
-
-      // Special handling to add duplicate topics to the response
-      def sendResponseWithDuplicatesCallback(results: Map[String, ApiError]): Unit = {
```
Are we no longer concerned about duplicate topics? Is that handled during deserialization?

bob-barrett reviewed (commented):
This looks good to me, Colin. I left a couple minor comments.

As a more general comment, I think we should include the docstrings in the schema files, so that we can eventually generate the protocol docs off of them. And if we're going to do that, we may actually want to be more precise about not changing field names from the current manual schemas, so that people don't get thrown off by changed names. I don't think either of these things necessarily has to happen in this patch, though.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreateTopicsResponse.java:
```diff
@@ -84,63 +44,43 @@
-        this.throttleTimeMs = throttleTimeMs;
-        this.errors = errors;
+    public CreateTopicsResponse(Struct struct, short version) {
+        this.data = new CreateTopicsResponseData(struct, version);
```
We moved away at some point from passing the version in the constructor. The idea was do the same for the request classes so that we could remove the builders altogether. What's the reasoning for reintroducing this?

cmccabe replied:
We need the version in order to parse the data that is in the `Struct` object.  Without the version, we don't know the format that we should expect.  Similarly, once this is converted to using ByteBuffers, we'll need the version before we can parse what is in the ByteBuffer.

The version is not stored in the response.  Responses should continue to represent the data, not a particular serialization of the data.  It would be nice if Request was the same way-- at some point we'll probably fix those classes to be like that too.

ijuma replied:
I see. This intermediate step is a bit weird since we use the `version` to create the `Struct` in the first place. Once we drop `Struct`, then it seems reasonable to have a `parse` method that takes a `ByteBuffer` and `version` and returns a `CreateTopicsResponse`.

cmccabe committed:
KAFKA-7609: Add Protocol Generator for Kafka

cmccabe commented:
> As a more general comment, I think we should include the docstrings in the schema files, so that we can eventually generate the protocol docs off of them.

That's a good point.  I added documentation to all JSON schema files, based on the original Struct documentation.  In cases where this documentation was missing, I created protocol documentation for the fields.  I also ported over the comments explaining the purpose of each protocol version change.

> And if we're going to do that, we may actually want to be more precise about not changing field names from the current manual schemas, so that people don't get thrown off by changed names. I don't think either of these things necessarily has to happen in this patch, though.

I think it's best to decide about field names when doing the conversion for each RPC.

> Are we no longer concerned about duplicate topics? Is that handled during deserialization?

I fixed this so that we handle duplicates the same way we do now: by setting an INVALID_REQUEST error for them.  This is actually specified in the KIP that added CreateTopics.

In order to make this possible, deserialization needs to be able to represent duplicate topics in the set.  I handled this by using multisets instead of sets for data received over the wire.  I think this is the correct approach in general because it's flexible and will allow us to covert over the existing code more easily.  Throwing a deserialization error when encountering duplicates would be very unfriendly-- deserialization errors just result in the broker disconnecting without an error message.

To make this easier to review, I have split the CreateTopics changes off into a separate PR, https://github.com/apache/kafka/pull/5972  .  I think with that stuff split out, this is a very low-impact change since it's just adding a code generator, and not modifying any existing broker code path.

bob-barrett commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java:
```diff
@@ -158,17 +189,20 @@ private static int slot(Element[] curElements, Element e) {
      */
-    private int findIndex(E example) {
-        int slot = slot(elements, example);
+    final private int findEqualElement(Object key) {
```
The method name sounds like it should return an element, not an index. Maybe change it to something like `findIndexOfEqualElement`?

cmccabe replied:
changed

bob-barrett commented on buildSrc/src/main/java/org/apache/kafka/message/CodeBuffer.java:
```diff
@@ -0,0 +1,85 @@
+    public int hashCode() {
+        return lines.hashCode();
+    }
+};
```
Why the trailing semicolons?

cmccabe replied:
Removed

bob-barrett reviewed (commented):
Thanks Colin! I agree about handling field names when we do the conversion. I added a couple more comments.

cmccabe commented:
* Rebase on trunk

* Update message schemas for KIP-380

cmccabe commented:
retest this please

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -0,0 +1,142 @@
+    public String about() {
+        return about;
+    }
+};
```
nit: unneeded semicolon (there are a few of these)

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/HeaderGenerator.java:
```diff
@@ -0,0 +1,78 @@
+        ""
+    };
+
+    static final String PACKAGE = "org.apache.kafka.message";
```
Maybe we should put this under `org.apache.kafka.common.message`?

cmccabe replied:
That makes sense.  I think we should have:
`org.apache.kafka.common.protocol`: Message, Readable, Writable, MessageUtil
`org.apache.kafka.message`: the generator code
`org.apache.kafka.common.message`: the generated message code

One of the reasons behind the split is that in newer versions of Java, you can only have one directory of sources that provides a certain Java module.

Another reason is because it makes it more clear what is providing what.  There is really no reason to have these three items in the same Java package since they provide distinct things.

hachikuji replied:
The suggested packages sound good to me.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/FieldType.java:
```diff
@@ -0,0 +1,281 @@
+            default:
+                if (string.startsWith(STRUCT_PREFIX)) {
+                    String elementTypeString = string.substring(STRUCT_PREFIX.length());
+                    if (string.length() == 0) {
```
I think you probably meant to use `elementTypeString` here.

cmccabe replied:
Good catch.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/CodeBuffer.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public void printf(String format, Object... args) throws IOException {
```
Seems `IOException` is not actually thrown.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/CodeBuffer.java:
```diff
@@ -0,0 +1,85 @@
+        this.indent = 0;
+    }
+
+    public int indent() {
```
Seems unused. Can we remove? Otherwise, maybe `currentIndent` is a better name? I initially expected this to mutate.

cmccabe replied:
Will remove

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/CodeBuffer.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public boolean equals(Object other) {
+        if ((other == null) || (!(other instanceof CodeBuffer))) {
```
nit: the null check is redundant.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageFactoryGenerator.java:
```diff
@@ -0,0 +1,125 @@
+
+    public void registerMessageType(MessageSpec spec) {
+        if (spec.type() == MessageSpecType.REQUEST) {
+            if (requestApis.containsKey(Short.valueOf(spec.apiKey()))) {
```
nit: the boxing seems unnecessary

cmccabe replied:
I never liked auto-boxing.  But for consistency's sake, I'll use it

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageFactoryGenerator.java:
```diff
@@ -0,0 +1,125 @@
+        buffer.printf("switch (apiKey) {%n");
+        buffer.incrementIndent();
+        for (Map.Entry<Short, String> entry : apis.entrySet()) {
+            buffer.printf("case %d:%n", entry.getKey());
```
I wonder if it would be useful to have an `enum` with all the request types and theirs ids.

cmccabe replied:
Yeah, I thought about that too.  However, I don't want to create a duplicate class that does the same thing as ApiKeys.  I think in the long term we should be able to automatically generate ApiKeys.

hachikuji commented on clients/src/main/resources/message/AddOffsetsToTxnRequest.json:
```diff
@@ -0,0 +1,32 @@
+// limitations under the License.
+
+{
+  "apiKey": 25,
```
The redundancy of some of this information is a little annoying. Have you considered merging the request and response schemas into one file? For example:
```json
{
  "apiKey": 25,
  "name": "AddOffsetsToTxn",
  "requestSchema": {},
  "responseSchema": {}
}
```

cmccabe replied:
I like the flexibility of being able to specify things that aren't requests or responses.  One example is the request header and response header.  There may be other stuff in the future.

Also, short files tend to be easier to understand than long files, for me at least.

Also, it's not really that redundant, is it?  The requests and responses don't really share much at all beyond the apiKey.  Well... I guess validVersions is the same too.  But the contents of each version will be totally different.

We could get clever with path names to avoid specifying "name" here, but it's probably not the right thing to do.  Remember there are still systems out there where filesystems are case insensistive and the FS can't tell the difference between `addoffsetstotxn.java` and `ADDOFFSETSTOTXN.JAVA` :disappointed:

hachikuji replied:
Yeah, there are tradeoffs. I find it useful to see both sides of the API in the same view, but it's just a matter of convenience. I thought the size seemed reasonable even if it included both request and response schemas. One benefit is that it makes the relation between the two schemas clear and we can get rid of the annoying comments like "Schema version is bumped for changes to the request schema." In practice, the schemas are changed rarely enough that it probably doesn't make too big of a difference either way.

cmccabe replied:
There is common stuff that will be added in the future, I think.  One example is some kind of definition of error code names-- that would go in its own file, and not be a request or a response.

hachikuji commented on clients/src/main/resources/message/CreatePartitionsResponse.json:
```diff
@@ -0,0 +1,35 @@
+      "about": "The partition creation results for each topic.", "fields": [
+      { "name": "Name", "type": "string", "versions": "0+",
+        "about": "The topic name." },
+      { "name": "ErrorCode", "type": "int16", "versions": "0+",
```
We have found it useful to document the possible error types inside the `XResponse` classes. Do you think that is suitable meta information to be included in the json schema somewhere?

cmccabe replied:
I guess the question is what you would expect to do with that information.  I guess you could throw an exception if someone tried to send an error that wasn't allowed by a given version.  That would be a good sanity check on the server side.

Probably the most difficult thing would be expressing the concept of versioning here in a readable way.  Having a separate entry each time a new allowable error is added (or removed) feels very unreadable.  We could perhaps have a map from values to versions like:
```
{
"0,1": "0+",
"2": "1+"
}
```
The other thing which we would probably want here is the ability to give names to the 0, 1, 2, etc. constants.  This is all doable but just would require a lot of work.

So, to make a long story short, I think we should consider this for the future.  In the meantime we are documenting this in comments in the JSON file.

hachikuji replied:
Primarily it is for documentation. Occasionally we will even bump a protocol version just to be able to return a new error code safely even if the schema is unchanged otherwise. It is useful to track this information _somewhere_ and I can't think of a better place than the schema.

Anyway, I'm fine leaving it for the future. We should create a JIRA though.

cmccabe replied:
Sounds good.  I created a follow-up JIRA here: https://issues.apache.org/jira/browse/KAFKA-7787

hachikuji commented on clients/src/test/java/org/apache/kafka/message/MessageTest.java:
```diff
@@ -0,0 +1,308 @@
+    @Test
+    public void testRoundTrips() throws Exception {
+        testMessageRoundTrips(new MetadataRequestData().setTopics(
+            Arrays.asList(new MetadataRequestData.MetadataRequestTopic[] {
```
nit: `asList` accepts varargs, so the array creation is redundant.

hachikuji commented on clients/src/test/java/org/apache/kafka/message/MessageTest.java:
```diff
@@ -0,0 +1,308 @@
+            } catch (UnsupportedVersionException e) {
+                throw new RuntimeException("No response message spec found for API " + apiKey);
+            }
+            if (apiKey.latestVersion() > message.highestSupportedVersion()) {
```
nit: why not turn these into assertions?

cmccabe replied:
I think `org.junit.Assert.fail` just turns around and throws an exception anyway.  But sure, `Assert.fail` might be more idiomatic.  Changed.

hachikuji replied:
What I meant is why don't we just use `assertEquals` or `assertTrue`?

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/StructSpec.java:
```diff
@@ -0,0 +1,70 @@
+                      @JsonProperty("versions") String versions,
+                      @JsonProperty("fields") List<FieldSpec> fields) {
+        this.name = Objects.requireNonNull(name);
+        this.versions = Versions.parse(versions, Versions.ALL);
```
nit: I found it a little unintuitive that `ALL` is the default. I'd have expected `NONE`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java:
```diff
@@ -0,0 +1,60 @@
+     * Get the length of the UTF8 representation of a string, without allocating
+     * a byte buffer for the string.
+     */
+    public static short utf8Length(CharSequence input) {
```
I assume the reason for copying this method was to try and isolate the dependencies which were used by the generated classes? Perhaps we add a javadoc comment to make that clear?

cmccabe replied:
Hmm.  I think it's OK for `common.protocol` to rely on `common.util`.  I will keep the functions separate to preserve the 16-bit length check, though.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageGenerator.java:
```diff
@@ -0,0 +1,200 @@
+/**
+ * The Kafka message generator.
+ */
+public final class MessageGenerator {
```
Not sure there are any great alternatives, but the overloading of "message" is slightly unfortunate. May be fine long term since the project is gradually preferring the term "record" for log messages.

cmccabe replied:
Yeah, I think we should start referring to "records" rather than "messages" in the context of the kafka log.  The "log message" vs. "RPC message" thing is an existing point of confusion and I was just using the existing terminology (unfortunately)

hachikuji commented on clients/src/main/resources/message/MetadataResponse.json:
```diff
@@ -0,0 +1,81 @@
+    ]},
+    { "name": "ClusterId", "type": "string", "nullableVersions": "2+", "versions": "2+", "ignorable": true,
+      "about": "The cluster ID that responding broker belongs to." },
+    { "name": "ControllerId", "type": "int32", "versions": "1+",
```
The default should be -1. I imagine there could be a lot of cases like this. For example, I think the default `clusterId` should be null. It may be hard to find these cases in an automated test since defaults are not packaged in the legacy schema types.

cmccabe replied:
Good point.  Fixed.

I agree that defaults are a tricky thing, and probably something we won't solve 100% until the conversions are done for each message.  Also, we don't currently support null as a default.  I think it's best to do that in a follow-on change.

hachikuji replied:
Yes, that's fine. Perhaps you can create JIRAs for the follow-ups we're identifying? The community may be able to help out with some guidance.

cmccabe replied:
I filed https://issues.apache.org/jira/browse/KAFKA-7788 as a follow-up.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageSpec.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public String generatedClassName() {
+        return struct.name() + "Data";
```
Given that these classes are all extensions of `Message`, I wonder if that should also be the suffix?

cmccabe replied:
The initial plan here is to make the existing hand-written `FooRequest` types contain `FooRequestData` fields as members.  That eases the transition.  So to do that it's better to avoid a name collision by generating `FooRequestData` instead of `FooRequest`, etc.

hachikuji replied:
Ok, makes sense.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashMultiSet.java:
```diff
@@ -0,0 +1,142 @@
+        }
+        int slot = slot(elements, key);
+        int bestSlot = INVALID_INDEX;
+        for (int seen = 0; seen < elements.length; seen++) {
```
nit: We're not using `seen`, so this could be replace with `foreach`. Same in `findAll`.

cmccabe replied:
I'm not sure `foreach` is really better here, since the variable that we use would be equally unused.  This loop itself is kind of a bit of paranoia on my part.  It prevents an infinite loop if the hash table is 100% full.  But the hash table should never get more than 50% full anyway...

hachikuji commented on clients/src/main/resources/message/FetchResponse.json:
```diff
@@ -0,0 +1,77 @@
+          { "name": "FirstOffset", "type": "int64", "versions": "4+",
+            "about": "The first offset in the aborted transaction." }
+        ]},
+        { "name": "Records", "type": "bytes", "versions": "0+", "nullableVersions": "0+",
```
I wonder if you have given any thought to how we would do sendfile. I imagine this will end up being the last API we convert.

cmccabe replied:
Yeah, I think this will be the last thing we convert.  It may end up requiring some custom types or code as well.  For example, in order to do an incremental send, we may use sub-classes directly rather than using the overarching FetchResponseData class, etc.

The most important thing here is managing buffering well.  I think sendfile itself is overrated since it is not compatible with a lot of features that are must-haves these days (security, compression, downconversion, etc.)

hachikuji commented on clients/src/main/resources/message/ListOffsetRequest.json:
```diff
@@ -0,0 +1,48 @@
+  "fields": [
+    { "name": "ReplicaId", "type": "int32", "versions": "0+",
+      "about": "The broker ID of the requestor, or -1 if this request is being made by a normal consumer." },
+    { "name": "IsolationLevel", "type": "int8", "versions": "2+",
```
I wonder if there should be an enum type in this schema. Looking at the generated code, this field is just represented as a byte, but it would be nice to have something friendlier to work with. Otherwise we'll just create the enum manually and do some conversion.

cmccabe replied:
Yeah, enums would be a good addition.  It's a bit trickier than it seems, though.  For one thing, we already have a lot of public/stable enum types, and we can't break them.  enums can't inherit from other classes, either, so that escape hatch is closed.

Perhaps a simple way to make it work would be generating simple constants
```
public static int NONE = 0;
public static int OFFSET_OUT_OF_RANGE = 1;
... etc. etc. ...
```

Then the existing public/stable enums could be retrofitted to grab the values from these declarations.  We could also generate classes in cases where there was no existing public thing.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    @JsonProperty("mapKey")
+    public boolean mapKey() {
```
nit: `isMapKey`?

cmccabe replied:
Can we leave this for now?  It would just be a very big change... and I thought we liked accessors that were the same as field names... :pray:

hachikuji replied:
The convention is a bit loose for boolean accessors. I find that `isXX` and `hasXX` are more readable, but it's a personal preference. We can leave it if you prefer.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashMultiSet.java:
```diff
@@ -0,0 +1,142 @@
+ * A memory-efficient hash multiset which tracks the order of insertion of elements.
+ * See org.apache.kafka.common.utils.ImplicitLinkedHashSet for implementation details.
+ *
+ * This class is a multi-set because it allows multiple elements to be inserted that are
```
Wondering if this should be the exceptional behavior and not the rule. Which APIs currently expect multiset semantics? I realize that there is a difficulty enforcing set semantics when parsing requests, but it might be nice to catch violations when building a response.

cmccabe replied:
I think we need multisets here in order to avoid breaking compatibility.  One example is CreateTopics -- the KIP specifies what should happen when there are duplicate elements, and it's not a disconnect.  I would expect a similar story for all of the RPCs, unfortunately.

hachikuji reviewed (commented):
Thanks, this looks great! I left a few minor comments and some questions.

cmccabe committed:
Move some namespaces

cmccabe committed:
Require struct versions to be set

cmccabe committed:
More style fixes and etc.

cmccabe committed:
Finish renaming

cmccabe committed:
Bump ListOffset for KIP-207

cmccabe commented:
The test failure was `kafka.api.CustomQuotaCallbackTest.testCustomQuotaCallback`, which is unrelated.

cmccabe committed:
Fix some javadoc, use assertTrue instead of Assert.fail in a case

cmccabe commented:
retest this please

hachikuji commented:
Mentioned offline, but we need to move the few test cases under `clients/src/test/java/org/apache/kafka/common/message/` to `clients/src/test/java/org/apache/kafka/message/`.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -0,0 +1,142 @@
+        this.ignorable = ignorable;
+        this.about = about == null ? "" : about;
+        if (!this.struct.fields().isEmpty()) {
+            if (!this.type.isArray()) {
```
Couldn't we make this check stricter? Would we ever expect the type to be something other than `STRUCT`?

cmccabe replied:
I refactored this code to clean it up a bit.  Not all `FieldSpecs` describe arrays-- only some of them do.  We should have a `toStruct()` method to make it convenient, but not embed a `StructSpec` unconditionally into the `FieldSpec`.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/FieldType.java:
```diff
@@ -0,0 +1,281 @@
+import java.util.Optional;
+
+public interface FieldType {
+    static final String STRUCT_PREFIX = "[]";
```
nit: `static final` is redundant since this is an interface

cmccabe replied:
Removed, thanks

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/HeaderGenerator.java:
```diff
@@ -0,0 +1,78 @@
+        this.imports.add(newImport);
+    }
+
+    public void generate() throws Exception {
```
nit: `Exception` not raised. A bunch of these in `MessageDataGenerator` and `MessageFactoryGenerator` as well. Probably fall out from removing the `throws` on `printf`.

cmccabe replied:
Good point.  I was able to remove quite a few of these.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -0,0 +1,1279 @@
+        return true;
+    }
+
+    private boolean generateNonNullCheck(Versions prevVersions, FieldSpec field) throws Exception {
```
This is unused.

cmccabe replied:
Removed

cmccabe committed:
Fix namespace for VersionsTest, MessageGeneratorTest

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Message.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Returns the API key of this message, or -1 if there is none.
+     */
+    short apiKey();
```
It's a little annoying that we have to expose this for all of the nested types. Have you considered having something like an `ApiMessage` which extends this?

cmccabe replied:
That is a fair point.  I will separate out Message as a subtype of the ApiMessage interface.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/MessageFactoryGenerator.java:
```diff
@@ -0,0 +1,125 @@
+
+    public void registerMessageType(MessageSpec spec) {
+        if (spec.type() == MessageSpecType.REQUEST) {
+            if (requestApis.containsKey(spec.apiKey())) {
```
Is it worth having any validation that `apiKey` is not -1? I was even considering whether `MessageSpec.apiKey` could return `Optional<Short>`.

cmccabe replied:
I can use `Optional` there.

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/StructSpec.java:
```diff
@@ -0,0 +1,74 @@
+        }
+        this.fields = Collections.unmodifiableList(fields == null ?
+            Collections.emptyList() : new ArrayList<>(fields));
+        this.hasKeys = fields == null ? false : fields.stream().anyMatch(f -> f.mapKey());
```
nit: `fields != null && fields.stream().anyMatch(f -> f.mapKey())`

hachikuji commented on buildSrc/src/main/java/org/apache/kafka/message/Versions.java:
```diff
@@ -0,0 +1,116 @@
+
+import java.util.Objects;
+
+public final class Versions {
```
Some high-level documentation would be helpful for this class. Maybe we can add a JIRA to write some documentation for the schema format?

cmccabe replied:
There is some description of how versions work in `./clients/src/main/resources/common/message/README.md`.  I will add some more to the JavaDoc of this class as well.

hachikuji reviewed (commented):
A few more small comments.

cmccabe committed:
Some syntax cleanups

cmccabe committed:
Separate ApiMessage and Message, add Optional, etc.

cmccabe committed:
StructSpec: stylistic improvement

cmccabe committed:
Versions: add JavaDoc

cmccabe committed:
Don't embed Struct into FieldSpec

hachikuji commented on clients/src/main/resources/common/message/RequestHeader.json:
```diff
@@ -14,6 +14,7 @@
 // limitations under the License.
 
 {
+  "apiKey": -1,
```
Why was this needed?

cmccabe replied:
It shouldn't be needed-- I'll remove it.

In the long term, we'll want to generate other things besides ApiMessage from JSON files (enums, common structures, etc.) but we can hold off on that for now.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiMessage.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * A Message which is part of the top-level Kafka API.
+ */
+public interface ApiMessage extends Message {
```
Would it make any sense to move the version fields into ApiMessage?

cmccabe replied:
Hmm... I think it makes sense to leave them in Message.  All Messages have minimum and maximum supported versions, not just ApiMessages.

cmccabe committed:
Cleanups

hachikuji commented:
retest this please

hachikuji reviewed (approved):
Thanks @cmccabe, LGTM. This is a major improvement. I'll merge after the builds complete.

cmccabe commented:
Failing test is `kafka.api.UserQuotaTest.testThrottledProducerConsumer`, which is definitely unrelated to this PR

junrao commented:
@cmccabe : I could find clients/src/main/message/README.md mentioned in the PR. Does that still exist?

cmccabe commented:
It was moved to `./clients/src/main/resources/common/message/README.md`

junrao commented on clients/src/main/resources/common/message/ListOffsetResponse.json:
```diff
@@ -0,0 +1,49 @@
+          "about": "The result offsets." },
+        { "name": "Timestamp", "type": "int64", "versions": "1+" },
+        { "name": "Offset", "type": "int64", "versions": "1+",
+          "about": "The timestamp associated with the returned offset." },
```
This comment is for Timestamp, not for Offset?

cmccabe replied:
Good catch.  I have filed a PR to fix this: https://github.com/apache/kafka/pull/6214
